### PR TITLE
ACS-4154: Add category count to relevant CRUD endpoints for categories

### PR DESCRIPTION
**Pull request by krdabrowski**
https://alfresco.atlassian.net/browse/ACS-4154

### CODE REVIEW

File: packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/UpdateCategoriesTests.java [hunk 1]
```diff
@@ -233,4 +233,27 @@ public void testUpdateCategory_withIgnoredInvalidIdInBodyAndExpect200()
         restClient.assertStatusCodeIs(OK);
         updatedCategory.assertThat().field(FIELD_NAME).is(categoryNewName);
     }
+
+    /**
+     * Check weather count present in update category request will be ignored.
```
Comment (tpage-alfresco) [inline packaging/tests/tas-restapi/src/test/java/org/alfresco/rest/categories/UpdateCategoriesTests.java:238]:
```suggestion
     * Check whether count present in update category request will be ignored.
```

**Review by tpage-alfresco — APPROVED** [review]
LGMT

File: repository/src/main/resources/alfresco/public-services-security-context.xml [hunk 2]
```diff
@@ -575,6 +575,7 @@
                 org.alfresco.service.cmr.search.CategoryService.deleteCategory=ACL_ALLOW
                 org.alfresco.service.cmr.search.CategoryService.getTopCategories=ACL_ALLOW
                 org.alfresco.service.cmr.search.CategoryService.getRootCategoryNodeRef=ACL_ALLOW
+                org.alfresco.service.cmr.search.CategoryService.getCategoriesCount=ACL_ALLOW
```
Comment (mpichura) [inline repository/src/main/resources/alfresco/public-services-security-context.xml:578]: Most likely - not needed - see my comment on `CategoryService` interface.

**Review by mpichura — COMMENTED** [review]
I think we don't need the new method in `CategoryService` to get categories count.
Anyway - most of my work in ACS-4160 would loose its point.

File: repository/src/main/java/org/alfresco/service/cmr/search/CategoryService.java [hunk 3]
```diff
@@ -259,4 +261,17 @@ default Optional<NodeRef> getRootCategoryNodeRef(final StoreRef storeRef)
...
+
+    /**
+     * Get categories by usage count. Result is a map of category IDs (short form - UUID) as key and usage count as value.
+     *
+     * @param storeRef Reference to node store.
+     * @return Map of categories IDs and usage count.
+     */
+    @Experimental
+    @Auditable(parameters = {"storeRef"})
+    default Map<String, Integer> getCategoriesCount(StoreRef storeRef)
```
[the remark this answers is not in the document]
Comment (krdabrowski) [reply repository/src/main/java/org/alfresco/service/cmr/search/CategoryService.java:273]: Ohh nice hint! I will test it.

[the remark this answers is not in the document]
Comment (krdabrowski) [reply repository/src/main/java/org/alfresco/service/cmr/search/CategoryService.java]: You're true! I've fixed it to use `CategoryService.getTopCategories()`

[the remark this answers is not in the document]
Comment (krdabrowski) [reply repository/src/main/java/org/alfresco/service/cmr/search/CategoryService.java]: You're right! I've tested it and it works fine - fix already pushed.

File: remote-api/src/main/java/org/alfresco/rest/api/Categories.java [hunk 4]
```diff
@@ -26,32 +26,60 @@
...
 import org.alfresco.service.Experimental;
+import org.alfresco.service.cmr.repository.StoreRef;
 
 @Experimental
 public interface Categories
 {
-    Category getCategoryById(String id, Parameters parameters);
+    Category getCategoryById(StoreRef storeRef, String id, boolean includeCount);
+
+    default Category getCategoryById(String id, boolean includeCount)
```
[the remark this answers is not in the document]
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/Categories.java:42]: I feel like those layers should be separated - I mean the REST framework logic from the service layer. IMO we should avoid using staff from `org.alfresco.rest.framework.resource.parameters` package like: `CollectionWithPagingInfo` or `Parameters` in service layer.

File: repository/src/main/java/org/alfresco/repo/search/impl/AbstractCategoryServiceImpl.java [hunk 5]
```diff
@@ -620,4 +623,38 @@ public Optional<NodeRef> getRootCategoryNodeRef(final StoreRef storeRef)
                 .map(ChildAssociationRef::getChildRef)
                 .findFirst();
     }
+
+    @Override
+    @Experimental
+    public Map<String, Integer> getCategoriesCount(final StoreRef storeRef)
```
[the remark this answers is not in the document]
Comment (krdabrowski) [reply repository/src/main/java/org/alfresco/repo/search/impl/AbstractCategoryServiceImpl.java]: Fixed

File: remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java [hunk 6]
```diff
@@ -95,63 +92,89 @@ public CategoriesImpl(AuthorityService authorityService, CategoryService categor
...
         {
             throw new InvalidArgumentException(NOT_A_VALID_CATEGORY, new String[]{id});
         }
 
         validateCategoryFields(fixedCategoryModel);
+        final Category category = mapToCategory(changeCategoryName(categoryNodeRef, fixedCategoryModel.getName()));
+
+        if (includeCount)
+        {
+            final Map<String, Integer> categoriesCount = categoryService.getCategoriesCount(storeRef);
```
[the remark this answers is not in the document]
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/CategoriesImpl.java]: Fixed

File: repository/src/main/resources/alfresco/public-services-security-context.xml (hunk 2 above)
Comment (krdabrowski) [reply repository/src/main/resources/alfresco/public-services-security-context.xml]: Fixed

File: remote-api/src/main/java/org/alfresco/rest/api/Categories.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mpichura) [reply remote-api/src/main/java/org/alfresco/rest/api/Categories.java:42]: The service layer is still in a subpackage of `org.alfresco.rest` (same as api framework layer). Avoiding these objects makes little sense until we do  a major refactor as many other service alike classes already use them.
Anyway, I can live with what is currently there but let's hope that we don't need to make interface changes when, for instance, someone wants to add sorting for get categories.

**Review by mpichura — APPROVED** [review]

File: remote-api/src/main/java/org/alfresco/rest/api/Categories.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/Categories.java]: I've rethink that and argument with categories sorting has convinced me - I've added the `Parameters` argument to methods.

**Pull request merged**